google/cloud/internal/rest_request.h

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_REST_REQUEST_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_REST_REQUEST_H
+
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <unordered_map>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+class RestClient;
+
+class RestRequest {
+ public:
+  using HttpHeaders = std::unordered_map<std::string, std::vector<std::string>>;
+  using HttpParameters = std::vector<std::pair<std::string, std::string>>;
+
+  RestRequest();
+  explicit RestRequest(std::string path);
+  RestRequest(std::string path, HttpHeaders headers);
+  RestRequest(std::string path, HttpParameters parameters);
+  RestRequest(std::string path, HttpHeaders headers, HttpParameters parameters);
+
+  RestRequest(RestRequest const&) = default;
```

Comment:
Do you need to spell these out?  I would have hoped the compiler generated them for you?